src/Endpoints/Indexes.php

```diff
@@ -0,0 +1,164 @@
+<?php
+
+namespace MeiliSearch\Endpoints;
+
+use MeiliSearch\Contracts\Endpoint;
+use MeiliSearch\Contracts\Http;
+use MeiliSearch\Endpoints\Delegates\handlesDocuments;
+use MeiliSearch\Endpoints\Delegates\handlesSettings;
+use MeiliSearch\Exceptions\TimeOutException;
+
+class Indexes extends Endpoint
+{
+    use handlesDocuments;
+    use handlesSettings;
+
+    const PATH = '/indexes';
+
+    /*
+     * @var string
+     */
+    private $uid;
+
+    public function __construct(Http $http, $uid = null)
+    {
+        $this->uid = $uid;
+        parent::__construct($http);
+    }
+
+    /**
+     * @param array $options
+     *
+     * @return $this
+     *
+     * @throws Exceptions\HTTPRequestException
+     */
+    public function create(string $uid, $options = [])
+    {
+        $options['uid'] = $uid;
+
+        $response = $this->http->post(self::PATH, $options);
+
+        return new self($this->http, $response['uid']);
+    }
+
+    public function all()
+    {
+        $indexes = [];
+
+        foreach ($this->http->get(self::PATH) as $index) {
+            $indexes[] = new self($this->http, $index['uid']);
+        }
+
+        return $indexes;
+    }
+
+    public function getPrimaryKey()
+    {
+        return $this->show()['primaryKey'];
+    }
+
+    public function getUid()
+    {
+        return $this->uid;
+    }
+
+    public function show()
+    {
+        return $this->http->get(self::PATH.'/'.$this->uid);
+    }
+
+    public function update($body)
+    {
+        return  $this->http->put(self::PATH.'/'.$this->uid, $body);
+    }
+
+    public function delete()
+    {
+        return $this->http->delete(self::PATH.'/'.$this->uid);
+    }
+
+    // Updates
+
+    public function getUpdateStatus($update_id)
+    {
+        return $this->http->get(self::PATH.'/'.$this->uid.'/updates/'.$update_id);
+    }
+
+    public function getAllUpdateStatus()
+    {
+        return $this->http->get(self::PATH.'/'.$this->uid.'/updates');
+    }
+
+    /**
+     * @param $update_id
+     * @param int $timeout_in_ms
+     * @param int $interval_in_ms
+     *
+     * @return mixed
+     *
+     * @throws TimeOutException
+     */
+    public function waitForPendingUpdate($update_id, $timeout_in_ms = 5000, $interval_in_ms = 50)
+    {
+        $timeout_temp = 0;
+        while ($timeout_in_ms > $timeout_temp) {
+            $res = $this->getUpdateStatus($update_id);
+            if ('enqueued' != $res['status']) {
+                return $res;
+            }
+            $timeout_temp += $interval_in_ms;
+            usleep(1000 * $interval_in_ms);
+        }
+        throw new TimeOutException();
+    }
+
+    // Search
+
+    public function search($query, array $options = [])
+    {
+        $parameters = array_merge(
+            ['q' => $query],
+            $this->parseOptions($options)
+        );
+
+        return $this->http->get(self::PATH.'/'.$this->uid.'/search', $parameters);
```

Comment:
Maybe we can create a value-object or a DTO which define a common structure for each search? 

```php
<?php

declare(strict_types=1);

namespace MeiliSearchBundle\Client;

final class Search implements SearchInterface
{
    /**
     * @var array<int,array>
     */
    private $hits;

    /**
     * @var int
     */
    private $offset;

    /**
     * @var int
     */
    private $limit;

    /**
     * @var int
     */
    private $nbHits;

    /**
     * @var int
     */
    private $exhaustiveNbHits;


    /**
     * @var int
     */
    private $processingTimeMs;

    /**
     * @var string
     */
    private $query;

    /**
     * {@inheritdoc}
     */
    public static function create(
        array $hits,
        int $offset,
        int $limit,
        int $nbHits,
        int $exhaustiveNbHits,
        int $processingTimeMs,
        string $query
    ): SearchInterface {
        $self = new self();

        $self->hits = $hits;
        $self->offset = $offset;
        $self->limit = $limit;
        $self->nbHits = $nbHits;
        $self->exhaustiveNbHits = $exhaustiveNbHits;
        $self->processingTimeMs = $processingTimeMs;
        $self->query = $query;

        return $self;
    }

    /**
     * {@inheritdoc}
     */
    public function filter(callable $callback): SearchInterface
    {
        $results = array_filter($this->hits, $callback, ARRAY_FILTER_USE_BOTH);

        return self::create($results, $this->offset, $this->limit, $this->nbHits, $this->exhaustiveNbHits, $this->processingTimeMs, $this->query);
    }

    /**
     * {@inheritdoc}
     */
    public function getHits(): array
    {
        return $this->hits;
    }

    /**
     * {@inheritdoc}
     */
    public function getOffset(): int
    {
        return $this->offset;
    }

    /**
     * {@inheritdoc}
     */
    public function getLimit(): int
    {
        return $this->limit;
    }

    /**
     * {@inheritdoc}
     */
    public function getNbHits(): int
    {
        return $this->nbHits;
    }

    /**
     * {@inheritdoc}
     */
    public function getExhaustiveNbHits(): int
    {
        return $this->exhaustiveNbHits;
    }

    /**
     * {@inheritdoc}
     */
    public function getProcessingTimeMs(): int
    {
        return $this->processingTimeMs;
    }

    /**
     * {@inheritdoc}
     */
    public function getQuery(): string
    {
        return $this->query;
    }
}

```

Here's a small example of what I've added in a WIP bundle, the idea is to define the core fields and add a `hits` one which is not tied to a specific structure 🤔

Reply:
I think this should be on a separate PR.

Reply:
Yes, I agree 🙂